### PR TITLE
separate implementation pkg from plugin pkg and add root_path in config

### PR DESCRIPTION
**Pull request by sedflix**
It solves a the zfs part of containerd/containerd#4514 and containerd/containerd#4513.

https://github.com/containerd/containerd/pull/4534 needs to merged along with this.

Signed-off-by: Siddharth Yadav <sedflix@gmail.com>

### CODE REVIEW

**Comment by sedflix** [comment]
Is the currently failing test due to my code or due to testing infra?

File: zfs.go [hunk 1]
```diff
@@ -66,7 +54,7 @@ func NewSnapshotter(root string) (snapshots.Snapshotter, error) {
 		return nil, err
 	}
 	if m.FSType != "zfs" {
-		return nil, errors.Wrapf(plugin.ErrSkipPlugin, "path %s must be a zfs filesystem to be used with the zfs snapshotter", root)
+		return nil, errors.New(fmt.Sprintf("path %s must be a zfs filesystem to be used with the zfs snapshotter", root))
```
Comment (AkihiroSuda) [inline zfs.go:57]: errors.Errorf

File: zfs.go [hunk 2]
```diff
@@ -20,12 +20,11 @@ package zfs
 
 import (
```
Comment (AkihiroSuda) [inline zfs.go:21]:
```
panic: sha256 not available (make sure it is imported) [recovered]
	panic: sha256 not available (make sure it is imported)
```

**Review by AkihiroSuda — APPROVED** [review]

File: zfs.go [hunk 3]
```diff
@@ -20,16 +20,14 @@ package zfs
...
 	"context"
-	"path/filepath"
-
+	_ "crypto/sha256"
```
Comment (dmcgowan) [inline zfs.go:23]: Why was this added?

Comment (sedflix) [reply zfs.go:23]: I was getting the following error before adding this import.

```
panic: sha256 not available (make sure it is imported) [recovered]
	panic: sha256 not available (make sure it is imported)
```

Comment (dmcgowan) [reply zfs.go:23]: Since this package does not directly use it, it is not the normally the place you would import it. Where was it importing from that ran into this error? You can import it from `_test.go` file if that is necessary.

Comment (thaJeztah) [reply zfs.go:23]: @sedflix ^^ could you have a look?

**Comment by thaJeztah** [comment]
Looks like you forgot the dco sign-off on the last commit; could you squash that commit with the previous one and fix the sign-off? Let me know if you need help doing so 🤗

```
28df8af - FAIL - does not have a valid DCO
3891 commits to fix
390The command "DCO_VERBOSITY=-q ../project/script/validate/dco" exited with 1.

**Comment by sedflix** [comment]
Yeah. I often forget to add `-s` since I rarely contribute here. 
This PR is turning out to be quite ugly 🙈 !

Thanks for pointing that out @thaJeztah !

File: zfs.go (hunk 3 above)
Comment (sedflix) [reply zfs.go]: @dmcgowan i shifted it to zfs_test.go. lmk if any other changes are required are here.

**Review by AkihiroSuda — APPROVED** [review]

**Review by estesp — APPROVED** [review]
LGTM
